### PR TITLE
build: support Poetry for depencencies tool in api's Dockerfile

### PR DESCRIPTION
**Pull request by bowenliang123**
# Description

- add a Dockerfile for building api Docker image, using Poetry for dependencies installation
- the same Debian dependencies as in existed api Dockerfile
- reusing the existed `entrypoint.sh` for entryscript
- add CI jobs for testing building API Docker image shipping with Poetry

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by bowenliang123** [comment]
It works.

image

**Comment by bowenliang123** [comment]
cc @takatost

**Review by takatost — CHANGES_REQUESTED** [review]
Can we push the API with the tag starting with "poetry-" to the image repository? (excluding the latest tag)
This way, we can use the built image to validate on both amd64 and arm64 hosts.

**Comment by bowenliang123** [comment]
> Can we push the API with the tag starting with "poetry-" to the image repository? (excluding the latest tag) This way, we can use the built image to validate on both amd64 and arm64 hosts.

Sorry that I failed to manage to do that. Getting the `DOCKERHUB_USER` and `DOCKERHUB_TOKEN` null from Github secrets.

I will suggest to have the poetry-based dockerfile in this PR , and do the image pushing in the follow-ups.

**Comment by bowenliang123** [comment]
https://github.com/milvus-io/milvus-lite/issues/137

**Comment by bowenliang123** [comment]
Remained blocking issue: `pymilvus` 2.4.x fails the dependencies installation on Linux arm64, as it requires `milvus-lite` which does not provide the prebuilt whl package for arm64 on Linux (referring to https://pypi.org/project/milvus-lite/2.4.7/#files).

Therefore, we may have to revert the PR #3795 ,  falling back to use `pymilvus` 2.3.x to avoid the blocker for applying Poery build system.

File: .github/workflows/api-tests.yml [hunk 1]
```diff
@@ -108,7 +108,7 @@ jobs:
 
       - name: Poetry check
         run: |
-          poetry check -C api
+          poetry check -C api --lock
```
Comment (bowenliang123) [inline .github/workflows/api-tests.yml:111]: declaimer: the `--lock` option does not change the lockfile , and it only makes sure the lockfile exists.

**Review by takatost — APPROVED** [review]
LGTM

**Comment by bowenliang123** [comment]
Most credit for this PR goes to @takatost for helping with the validation and refactoring of a runnable version of GitHub workflow for building and pushing the docker images. Thanks.

**Comment by junjiejiangjjj** [comment]
Milvus-lite 2.4.8 now supports poetry and arm64

**Pull request merged**